### ISSUE TITLE
How could we control Random State?

### ISSUE THREAD

**Issue by yuchenlin** [guillaumegenthial/sequence_tagging#6]
Hi, Guillaume,

I am very confused about why the training result is quite random since it seems that there is no shuffling of the data in the code.

I tried to set `tf.set_random_seed(42)` and `np.random.seed(42)` at the beginning of model.py, but it does not work.

Thanks very much,
Bill

**Comment by yuchenlin** [comment]
Additional Info: The result of the first epoch is always the same.

**Comment by yuchenlin** [comment]
OIC.. the random effect seems to be due to the Dropout.

**Comment by guillaumegenthial** [comment]
You're right, there is no shuffling anywhere. Randomness surely comes from dropout + initialization. However, if you carefully control the random seeds, you should get non-random convergence!

**Comment by yuchenlin** [comment]
Thanks so much. although simply setting the random state does not work, I found a page [here](https://stackoverflow.com/questions/42156296/dropoutwrapper-being-non-deterministic-across-runs) regarding the exactly the same question. 
I will have a try!

**Comment by loveJasmine** [comment]
maybe you should also set python's random seed fixed

**Comment by yuchenlin** [comment]
It seems that even if I set dropout=1, the results are still changing.

**Comment by yuchenlin** [comment]
@loveJasmine  i have tried but not effective...: (

**Comment by yuchenlin** [comment]
It seems that setting random seed is effective to some extent but cannot keep the results 100% exactly the same. I guess it maybe due to the multi-thred thing.

It seems that the tensorflow is often complained by users about the random state stuff..

**Comment by guillaumegenthial** [comment]
Have you tried running it on CPU on a subset of the dataset with a batch of size 1 with dropout = 1? I'm not too familiar with tf random seeds, and the stackoverflow post you refered above seems to mention some pretty weird behavior. Batch size = 1 could alleviate the problem they're mentioning.

**Comment by grafael** [comment]
@yuchenlin maybe this help: https://stackoverflow.com/questions/38469632/tensorflow-non-repeatable-results

**Comment by guillaumegenthial** [comment]
@yuchenlin did you make progress on this issue? Should we keep it open or should we close it?

**Comment by yuchenlin** [comment]
@guillaumegenthial  sorry for the delayed reply! I was busy doing other stuff. Yes, I guess we can close it. It seems that the current methods are difficult to maintain the stable results under the GPU settings. Thanks!

**Comment by guillaumegenthial** [comment]
@yuchenlin no problem! I'll close it, but I must confess that this is an interesting issue and I'd love to hear more about it if you find the answer one day.
